src/vos/tests/vts_pm.c

```diff
@@ -1943,6 +1943,284 @@ test_inprogress_parent_punch(void **state)
+	int			total = 0, success = 0, writes = 0;
+
+	test_args_reset(arg, VPOOL_SIZE);
+	coh = arg->ctx.tc_co_hdl;
+
+	memset(&iod, 0, sizeof(iod));
+
+	rc = daos_sgl_init(&sgl, 1);
+	assert_int_equal(rc, 0);
+	rc = daos_sgl_init(&fetch_sgl, 1);
+	assert_int_equal(rc, 0);
+
+	/* Set up dkey and akey */
+	for (i = 0; i < NR_OBJ; i++)
+		oid[i] = gen_oid(arg->ofeat);
+	for (i = 0; i < NR_DKEY; i++) {
+		vts_key_gen(&dkey_buf[i][0], arg->dkey_size, true, arg);
+		set_iov(&dkey[i], &dkey_buf[i][0],
+			arg->ofeat & DAOS_OF_DKEY_UINT64);
+	}
+	for (i = 0; i < NR_AKEY; i++) {
+		vts_key_gen(&akey_buf[i][0], arg->akey_size, true, arg);
+		set_iov(&akey[i], &akey_buf[i][0],
+			arg->ofeat & DAOS_OF_AKEY_UINT64);
+		iod[i].iod_type = DAOS_IOD_SINGLE;
+		iod[i].iod_size = strlen(first);
+		iod[i].iod_name = akey[i];
+		iod[i].iod_recxs = NULL;
+		iod[i].iod_nr = 1;
+	}
+	d_iov_set(&sgl.sg_iovs[0], (void *)first, iod[0].iod_size);
+	d_iov_set(&fetch_sgl.sg_iovs[0], (void *)buf, sizeof(buf));
+
+	tx_num = 0;
+start_over:
+	srand(0);
+	for (i = 0; i < NR_OBJ; i++) {
+		for (j = 0; j < NR_DKEY; j++) {
+			for (k = 0; k < NR_AKEY; k++) {
+				for (count = 0; count < 3; count++) {
+					total++;
+					switch (tx_num & 3) {
+					case 0:
+						epoch -= 3;
+						break;
+					case 1:
+						epoch += 2;
+						break;
+					case 2:
+						epoch -= 1;
+						break;
+					case 3:
+						epoch += 10;
+						break;
+					}
+
+					cur_tx = tx_num++ % NR_TX;
+					old_tx = (cur_tx + 1) % NR_TX;
+					vts_dtx_begin_ex(&oid[i], coh,
+							 epoch, 0, 0, 1,
+							 &req[cur_tx].dth);
+					req[cur_tx].oid = oid[i];
+					req[cur_tx].coh = coh;
+					req[cur_tx].xid =
+						req[cur_tx].dth->dth_xid;
+
+					random = rand();
+					flags = 0;
+					op = random % TX_NUM_OPS;
+					switch (op) {
+					case TX_OP_PUNCH_OBJ:
+						do_punch(&req[cur_tx]);
+						break;
+					case TX_OP_COND_PUNCH_DKEY:
+						flags = DAOS_COND_PUNCH;
+						req[cur_tx].flags = flags;
+					case TX_OP_PUNCH_DKEY:
+						req[cur_tx].dkey = &dkey[j];
+						do_punch(&req[cur_tx]);
+						break;
+					case TX_OP_COND_PUNCH_AKEY:
+						flags = DAOS_COND_PUNCH;
+						req[cur_tx].flags = flags;
+					case TX_OP_PUNCH_AKEY:
+						req[cur_tx].dkey = &dkey[j];
+						req[cur_tx].akey = &akey[k];
+						req[cur_tx].akey_nr = 1;
+						do_punch(&req[cur_tx]);
+						break;
+					default:
+						req[cur_tx].dkey = &dkey[j];
+						req[cur_tx].iod = &iod[k];
+						req[cur_tx].sgl = &sgl;
+						req[cur_tx].fetch_sgl =
+							&fetch_sgl;
+						req[cur_tx].akey_nr = 1;
+						do_io(&req[cur_tx], op);
+						break;
+					}
+					if (req[cur_tx].success)
+						success++;
+					if (req[cur_tx].commit)
+						writes++;
+					vts_dtx_end(req[cur_tx].dth);
+					if (req[old_tx].commit) {
+						rc = vos_dtx_commit(coh,
+							    &req[old_tx].xid, 1,
+							    NULL);
+						assert_int_equal(rc, 1);
+					}
+					memset(&req[old_tx], 0, sizeof(req[0]));
+				}
+			}
+		}
+		if ((epoch - 200) < epr.epr_lo)
```

Comment:
just curious, why - 200?

Reply:
yeah, the idea is avoid aggregating anything still in a transaction.  Now that I think about it, maybe it should be higher than that.   I'll fix in a separate patch unless I update this one again.